src/features/feature_selector.cpp

```diff
@@ -0,0 +1,241 @@
+#include "index/postings_data.h"
+#include "io/binary.h"
+
+namespace meta
+{
+namespace features
+{
+feature_selector::feature_selector(const std::string& prefix,
+                                   std::shared_ptr<index::forward_index> idx)
+    : prefix_{prefix},
+      idx_{std::move(idx)},
+      selected_{prefix_ + ".selected", idx_->unique_terms()}
+{ /* nothing */
+}
+
+void feature_selector::init(uint64_t features_per_class)
+{
+    // if the first class distribution doesn't exist, we haven't created the
+    // data for this feature_selector yet
+    if (!filesystem::file_exists(prefix_ + ".1"))
+    {
+        // initially set all probabilities to zero; this allows fast random
+        // access to the probabilities
+        term_prob_.assign(idx_->unique_terms(), 0.0);
+        class_prob_.assign(idx_->num_labels(), 0.0);
+        co_occur_.assign(idx_->num_labels(),
+                         std::vector<double>(idx_->unique_terms(), 0.0));
+        calc_probs();
+        score_all();
+        select(features_per_class);
+    }
+}
+
+void feature_selector::score_all()
+{
+    using pair_t = std::pair<term_id, double>;
+    std::vector<std::vector<pair_t>> scores(
+        class_prob_.size(), std::vector<pair_t>(term_prob_.size()));
+
+    printing::progress prog{" > Selecting features: ", term_prob_.size()};
+    for (uint64_t tid = 0; tid < idx_->unique_terms(); ++tid)
+    {
+        prog(tid);
+        for (uint64_t lbl = 0; lbl < idx_->num_labels(); ++lbl)
+            scores[lbl][tid] = std::make_pair(
+                tid, score(static_cast<label_id>(lbl + 1), term_id{tid}));
+    }
+    prog.end();
+
+    parallel::parallel_for(
+        scores.begin(), scores.end(), [&](std::vector<pair_t>& v)
+        {
+            std::sort(v.begin(), v.end(), [&](const pair_t& a, const pair_t& b)
+                      {
+                          return a.second > b.second;
+                      });
+        });
+
+    for (uint64_t lbl = 0; lbl < idx_->num_labels(); ++lbl)
+    {
+        // write (term_id, score) pairs
+        std::ofstream out{prefix_ + "." + std::to_string(lbl + 1)};
+        for (auto& score : scores[lbl])
+        {
+            io::write_binary(out, score.first);
+            io::write_binary(out, score.second);
+        }
+    }
+}
+
+void feature_selector::select(uint64_t features_per_class /* = 20 */)
+{
+    term_id id;
+    double score;
+    std::unordered_set<term_id> terms;
+    for (uint64_t lbl = 0; lbl < idx_->num_labels(); ++lbl)
+    {
+        std::ifstream in{prefix_ + "." + std::to_string(lbl + 1)};
+        for (uint64_t i = 0; i < features_per_class; ++i)
+        {
+            io::read_binary(in, id);
+            io::read_binary(in, score);
+            terms.insert(id);
+        }
+    }
+
+    // zero out old vector
+    for (auto& b : selected_)
+        b = false;
+
+    // select new features
+    for (auto& term : terms)
+        selected_[term] = true;
```

Comment:
If this is all `terms` is used for, is there any reason to make the temporary at all? Zero as the first step, then change line 95 to `selected_[id] = true;`?